dataflow/src/main/java/org/checkerframework/dataflow/util/PurityUtils.java

```diff
@@ -38,12 +39,39 @@ public static boolean isDeterministic(AnnotationProvider provider, MethodTree tr
         return isDeterministic(provider, methodElement);
     }
 
+    /**
+     * Is the method {@code tree} deterministic?
+     *
+     * <p>This implementation does not require an AnnotationProvider and does not handle aliases.
+     * There are currently no aliases for @Deterministic, and the only alias for @Pure is
```

Comment:
I think it would be much nicer to not add these additional overloaded versions.
The single use of these methods in `FlowExpressions.java` can be changed to use the available `AnnotationProvider`.
The issue isn't just aliases that this version would miss, it is also stub file support. Without an AnnotationProvider, this logic will not support stub files. This will lead to hard to debug problems and strange behavior.

Reply:
> The single use of these methods in `FlowExpressions.java` can be changed to use the available `AnnotationProvider`.

What is the available `AnnotationProvider`?  I don't see one that is available at the call to `isDeterministic`, but maybe I missed something.

Reply:
I think this is actually even easier: note that we only create a `MethodCall` if the method is deterministic, as checked here: https://github.com/typetools/checker-framework/blob/master/dataflow/src/main/java/org/checkerframework/dataflow/analysis/FlowExpressions.java#L176

https://github.com/typetools/checker-framework/pull/2384/commits/165101757681cd655fd424defbef4d9d0ecebe10 undoes these changes and locally tests passed.
Does this look good?

Reply:
Clever!

This subtle property definitely needs to be documented, both at the call site that you removed and in the documentation for MethodCall.

Reply:
I added the documentation for the change.